ONSdigital/census-contact-centre-service
# Cr 67

# Motivation and Context

This change is required because of CR-67, which states:

POST /cases/{caseid}/fulfilment/post
The service used by the endpoint should perform some business validation against the request to ensure that :
a. the product for the fulfilmentCode exists! b. The deliveryChannel for the requested product matches that of the endpoint ie POST c. The fulfilmentCode being requested is applicable to the region that the household case is in - this will entail calling the case service to retrieve the case details, so is dependent on CR-86 being completed first.
The product details will need retrieving from the ProductReference class - see the GET /fulfilments endpoint code to see how it gets product details, in particular the service code it uses - you will need to create an example Product and populate the fulfilmentCode and the requestChannel="CC" and deliveryChannel="POST" to find the qualifying Product.
When the service signals to the endpoint that one of the above tests fails, the endpoint should return to the client a BAD_REQUEST
In order to determine the region for the household case in hand, the service will need to retrieve the case details from the RM case service - this will be done using a RestClient implementation to call Case service (needed by CR-86)

# What has changed
The following class has been changed to include methods that will receive an incoming PostalFulfilmentRequestDTO, from the Contact Centre, then use to create an outgoing FulfilmentRequestedEvent, to send to RM. The FulfilmentRequestedEvent gets put onto the events exchange, which in turn puts it onto an outgoing queue named Fulfilment.Request. 

The incoming JSON, for the PostalFulfilmentRequestDTO, would look something like this:

{
 "caseId": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
 "title": "Mr",
 "forename": "Mickey",
 "surname": "Mouse",
 "fulfilmentCode": "P_OR_I1",
 "dateTime": "2016-11-09T11:44:44.797Z"
}

The outgoing JSON, for the FulfilmentRequestedEvent, would look something like this:

{"event":
	{"type":"FULFILMENT_REQUESTED",
	"source":"CONTACT_CENTRE_API",
	"channel":"CC",
	"dateTime":"2019-05-21T15:55:16.473Z",
	"transactionId":"28695b5f-aa1e-40f4-89ca-de0d08c1d8e2"},
		"payload":
		{"fulfilmentRequest":
			{"fulfilmentCode":"P_OR_I1",
			"caseId":"3fa85f64-5717-4562-b3fc-2c963f66afa6",
			"individualCaseId":"c56a52cd-fead-49ce-baae-4019bcd1e3a0",
			"address":		 {"addressLine1":null,"addressLine2":null,"addressLine3":null,"townName":null,"postcode":null,"region":null,"latitude":null,"longitude":null,"uprn":null,"arid":null,"addressType":null,"estabType":null},
		"contact":
		{"title":"Mr","forename":"Mickey","surname":"Mouse","email":null,"telNo":null}}}}





# How to test?
Do as follows:
Add Peter’s class, named CaseServiceFakeStub into the following package:

uk.gov.ons.ctp.integration.contactcentresvc.endpoint

Amend the application.yml so that it contains the following for the case service settings:
case-service-settings:
  case-by-id-query-path: /cases-rm-fake/{case-id}
  case-by-uprn-query-path: /cases-rm-fake/uprn/{uprn}
  case-by-case-reference-query-path: /cases-rm-fake/ref/{reference}
  rest-client-config:
    username: serco_cks
    password: temporary
    scheme: http
    host: localhost
    port: 8171

Run the following command to launch RabbitMQ from a terminal (in the census-contact-centre-service repo):

Docker-compose up -d

Save, compile and run the Contact Centre service. 

Run Postman and use the example JSON above to post a PostalFulfilmentRequestDT to the following endpoint:

http://localhost:8171/cases/3fa85f64-5717-4562-b3fc-2c963f66afa6/fulfilment/post

A FulfilmentRequestedEvent message should then appear in the Fulfilment.Request queue. NB. The Fulfilment.Request queue has now been removed from the code (as it should be created by the consumer service instead) and so for testing you will need to create it manually, using the RabbitMQ management console, and manually bind it to the events exchange e.g. using the binding event.fulfilment.request
